### PR TITLE
Program analysis

### PR DESCRIPTION
**Pull request by RamIowaStateUniversity**
- Adding traversal and fixpoint type, statement and expression
- Adding visitors to added types, statements and expressions
- Modified grammar files to accommodate traversal and fixp. 

I will change fixpoint to normal function after merging the program analysis feature (I will add this too to TODO list). In the next request, I will also add a test boa file for traversal feature.

### CODE REVIEW

**Comment by psybers** [comment]
The file BoaNoActions.g also needs updated.  This is the same as Boa.g but without the semantic actions - thus making it easier to read the grammar.

File: src/java/boa/compiler/SymbolTable.java [hunk 1]
```diff
@@ -128,6 +129,10 @@
 			new StatementKindProtoMap(),
 			new TypeKindProtoMap(),
 			new VisibilityProtoMap(),
+			new CFGNodeTypeProtoMap(),
```
Comment (psybers) [inline src/java/boa/compiler/SymbolTable.java]: Minor issue: we keep this list alphabetized.

File: src/java/boa/compiler/ast/statements/FixPStatement.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+ * @author rramu
+ */
+public class FixPStatement extends Statement {
+	protected Component param1;
+	protected Component param2;
+	protected final List<Identifier> ids = new ArrayList<Identifier>();
+	protected Block body;
+	protected Expression condition;
+	public AbstractType returnType;
+	protected final List<Component> args = new ArrayList<Component>();
```
Comment (psybers) [inline src/java/boa/compiler/ast/statements/FixPStatement.java]: args isnt used

File: src/java/boa/compiler/ast/statements/TraverseStatement.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+ */
+public class TraverseStatement extends Statement {
+	protected boolean wildcard = false;
+	protected Component node;
+	protected final List<Identifier> ids = new ArrayList<Identifier>();
+	protected Block body;
+	protected Expression condition;
+	protected List<IfStatement> ifStatements = new ArrayList<IfStatement>();
+	public AbstractType returnType;
+	protected final List<Component> args = new ArrayList<Component>();
```
Comment (psybers) [inline src/java/boa/compiler/ast/statements/TraverseStatement.java]: args isnt used

File: src/java/boa/compiler/visitors/CodeGeneratingVisitor.java [hunk 4]
```diff
@@ -1668,6 +1817,24 @@ public void visit(final FunctionType n) {
 
 	/** {@inheritDoc} */
 	@Override
+	public void visit(final FixPType n) {
+		if (!(n.type instanceof BoaFunction))
```
Comment (psybers) [inline src/java/boa/compiler/visitors/CodeGeneratingVisitor.java]: this check isnt needed in codegen - if it passed typecheck it should be fine here

File: src/java/boa/compiler/visitors/TypeCheckingVisitor.java [hunk 5]
```diff
@@ -88,6 +89,90 @@ public void visit(final VisitStatement n) {
 		}
 	}
 
+	protected class TraversalCheckingVisitor extends AbstractVisitorNoArg {
+		protected Set<String> befores = new HashSet<String>();
```
Comment (psybers) [inline src/java/boa/compiler/visitors/TypeCheckingVisitor.java]: befores/afters/whens dont seem to really be used?

File: src/java/boa/compiler/visitors/TypeCheckingVisitor.java [hunk 6]
```diff
@@ -88,6 +89,90 @@ public void visit(final VisitStatement n) {
...
+			if (nested)
+				return;
+			nested = true;
+			n.getBody().accept(this);
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void visit(final FixPStatement n) {
+			final Set<String> s = befores;
```
Comment (psybers) [inline src/java/boa/compiler/visitors/TypeCheckingVisitor.java]: no need for a local - just use the field below

File: src/java/boa/compiler/visitors/TypeCheckingVisitor.java [hunk 7]
```diff
@@ -1065,6 +1233,33 @@ public void visit(final VisitorExpression n, final SymbolTable env) {
...
+		n.type = n.getType().type;
+	}
+
+	/** {@inheritDoc} */
+	@Override
+	public void visit(final FixPExpression n, final SymbolTable env) {
+		n.env = env;
+		n.getType().accept(this, env);
+		for (final Statement s : n.getBody().getStatements())
+			if (!(s instanceof FixPStatement))
```
Comment (psybers) [inline src/java/boa/compiler/visitors/TypeCheckingVisitor.java]: should this be enforced here? or by the parser?

**Comment by psybers** [comment]
[junit] ------------- Standard Error -----------------
    [junit] BoaJava.stg 168:44: \n in string
    [junit] BoaJava.stg 169:44: \n in string
    [junit] BoaJava.stg 170:1: \n in string
    [junit] BoaJava.stg 171:25: \n in string
    [junit] BoaJava.stg 172:21: \n in string
    [junit] BoaJava.stg 173:8: \n in string
    [junit] BoaJava.stg 174:22: \n in string
    [junit] BoaJava.stg 177:39: \n in string
    [junit] BoaJava.stg 178:33: \n in string
    [junit] BoaJava.stg 179:1: \n in string
    [junit] BoaJava.stg 180:25: \n in string
    [junit] BoaJava.stg 181:21: \n in string
    [junit] BoaJava.stg 182:8: \n in string
    [junit] BoaJava.stg 183:22: \n in string
    [junit] BoaJava.stg 228:40: \n in string
    [junit] BoaJava.stg 229:83: \n in string
    [junit] BoaJava.stg 230:8: \n in string
    [junit] BoaJava.stg 231:1: \n in string
    [junit] BoaJava.stg 232:0: \n in string
    [junit] BoaJava.stg 233:9: \n in string
    [junit] BoaJava.stg 234:69: \n in string
    [junit] BoaJava.stg 235:49: \n in string
    [junit] BoaJava.stg 236:1: \n in string
    [junit] ------------- ---------------- ---------------

**Comment by psybers** [comment]
Also I believe your changes to the Boa.g file have introduced some ambiguities into the parser, causing it to take 50-100% longer (4/3 seconds vs 2 seconds).

**Comment by RamIowaStateUniversity** [comment]
Thanks Dr.Dyer.

I resolved the \n problem in stg file. This will be a part of the next
request. I will also try to resolve the ambiguities when I make the next
request.

Thanks,
Ram

**Comment by RamIowaStateUniversity** [comment]
Dr.Dyer,

Shall I merge the pull request, so that I can make further commits?

Thanks,
Ram

**Pull request merged**